src/nvim/buffer.c

```diff
@@ -3272,212 +3433,362 @@ build_stl_str_hl (
-      prevchar_isitem = TRUE;
-      t = nstr;
+      prevchar_isitem = true;
+
+      // { Build the formatting string
+      char_u nstr[20];
+      char_u *t = nstr;
       if (opt == STL_VIRTCOL_ALT) {
         *t++ = '-';
         minwid--;
       }
       *t++ = '%';
-      if (zeropad)
+      if (zeropad) {
         *t++ = '0';
+      }
+
+      // Note: The `*` means we take the width as one of the arguments
       *t++ = '*';
-      *t++ = nbase == 16 ? base : (char_u)(nbase == 8 ? 'o' : 'd');
+      *t++ = (char_u) (base == kNumBaseHexadecimal ? 'X'
+                        : (base == kNumBaseOctal ? 'o'
+                        : 'd'));
       *t = 0;
+      // }
+
+      // { Determine how many characters the number will take up when printed
+      //  Note: We have to cast the base because the compiler uses
+      //        unsigned ints for the enum values.
+      long num_chars = 1;
+      for (long n = num; n >= (int) base; n /= (int) base) {
+        num_chars++;
+      }
 
-      for (n = num, l = 1; n >= nbase; n /= nbase)
-        l++;
-      if (opt == STL_VIRTCOL_ALT)
-        l++;
-      if (l > maxwid) {
-        l += 2;
-        n = l - maxwid;
-        while (l-- > maxwid)
-          num /= nbase;
+      // VIRTCOL_ALT takes up an extra character because
+      // of the `-` we added above.
+      if (opt == STL_VIRTCOL_ALT) {
+        num_chars++;
+      }
+      // }
+
+      size_t remaining_buf_len = (out_end_p - out_p) + 1;
+
+      // If the number is going to take up too much room
+      // Figure out the approximate number in "scientific" type notation.
+      // Ex: 14532 with maxwid of 4 -> '14>3'
+      if (num_chars > maxwid) {
+        // Add two to the width because the power piece will take
+        // two extra characters
+        num_chars += 2;
+
+        // How many extra characters there are
+        long n = num_chars - maxwid;
+
+        // { Reduce the number by base^n
+        while (num_chars-- > maxwid) {
+          num /= base;
+        }
+        // }
+
+        // { Add the format string for the exponent bit
         *t++ = '>';
         *t++ = '%';
+        // Use the same base as the first number
         *t = t[-3];
         *++t = 0;
-        vim_snprintf((char *)p, outlen - (p - out), (char *)nstr,
+        // }
+
+        vim_snprintf((char *)out_p, remaining_buf_len, (char *)nstr,
             0, num, n);
-      } else
-        vim_snprintf((char *)p, outlen - (p - out), (char *)nstr,
+      } else {
+        vim_snprintf((char *)out_p, remaining_buf_len, (char *)nstr,
             minwid, num);
-      p += STRLEN(p);
-    } else
+      }
+
+      // Advance the output buffer position to the end of the
+      // number we just printed
+      out_p += STRLEN(out_p);
+
+    // Otherwise, there was nothing to print so mark the item as empty
+    } else {
       item[curitem].type = Empty;
+    }
 
-    if (opt == STL_VIM_EXPR)
+    // Free the string buffer if we allocated it.
```

Comment:
I changed the wording a little. 
I mainly want to leave this comment in because the `if()` statement is checking the one place where we did allocate the string buffer. The comment makes it more obvious why we are doing the deallocation only in that case (or potentially other cases in the future).